File: app/src/main/java/org/zeroxlab/imgorg/lib/StoreOperation.java
// vim: et sw=4 sts=4 tabstop=4
package org.zeroxlab.imgorg.lib;

import android.content.ContentValues;
import android.content.Context;
import android.provider.MediaStore;
import android.util.Log;

import java.io.File;
import java.text.SimpleDateFormat;
import java.util.Date;

public class StoreOperation implements Operation {

    private final static SimpleDateFormat sSDF = new SimpleDateFormat("yyyy-MM");

    private String mSource;
    private String mDestination;
    private Date mDate;
    private boolean mFinished;

    StoreOperation(Media source, String dest) {
        mSource = source.data;
        mDate = source.date;
        Date d = mDate == null ? new Date() : mDate;
        String dstDate = sSDF.format(d);
        File dst = new File(dest + "/" + dstDate);
        mDestination = dst.getAbsolutePath();
    }

    @Override
    public void consume(Context ctx) {
        ContentValues values = new ContentValues();
        values.put(MediaStore.MediaColumns.DATA, mDestination);
        int rows = ctx.getContentResolver().update(
                MediaStore.Images.Media.EXTERNAL_CONTENT_URI,
                values,
                MediaStore.MediaColumns.DATA + "='" + mSource + "'",
                null
        );
        if (rows == 1) {
            File from = new File(mSource);
            File to = new File(mDestination);
            File parent = to.getParentFile();
            if (!parent.exists()) {
                parent.mkdirs();
            }
            from.renameTo(to);
        }
        mFinished = true;
    }

    @Override
    public boolean isFinished() {
        return mFinished;
    }

    @Override
    public String getSource() {
        return mSource;
    }

    @Override
    public String getDestination() {
        return mDestination;
    }
}

File: app/src/main/java/cc/jchu/imgorg/lib/StoreOperation.kt
// vim: et sw=4 sts=4 tabstop=4
package cc.jchu.imgorg.lib

import android.content.ContentValues
import android.content.Context
import android.provider.MediaStore
import cc.jchu.imgorg.lib.StoreOperation
import java.io.File
import java.text.SimpleDateFormat
import java.util.Date

class StoreOperation internal constructor(source: Media, dest: String) : Operation {

    override val source: String?
    override val destination: String
    private val mDate: Date?
    override var isFinished = false
        private set

    init {
        this.source = source.data
        mDate = source.date
        val dstDate = sSDF.format(mDate)
        val src = File(this.source)
        val dst = File(dest + "/" + dstDate + "/" + src.name)
        destination = dst.absolutePath
    }

    override fun consume(ctx: Context?) {
        val values = ContentValues()
        values.put(MediaStore.Images.Media.DATA, destination)
        val rows = ctx!!.contentResolver.update(
            MediaStore.Images.Media.EXTERNAL_CONTENT_URI,
            values,
            MediaStore.Images.Media.DATA + "='" + source + "'",
            null
        )
        if (rows == 1) {
            val from = File(source)
            val to = File(destination)
            val parent = to.parentFile
            if (!parent.exists()) {
                parent.mkdirs()
            }
            from.renameTo(to)
        }
        isFinished = true
    }

    companion object {
        private val sSDF = SimpleDateFormat("yyyy-MM")
    }
}
